soc/st/stm32/common/stm32_bootloader.c

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2020 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <zephyr/init.h>
+#include <zephyr/retention/retention.h>
+#include <zephyr/retention/bootmode.h>
+#include <stm32_ll_system.h>
+
+static const uint32_t bootloader = DT_REG_ADDR(DT_INST(0, st_stm32_bootloader));
+static FUNC_NORETURN void jump_to_bootloader(void)
+{
+	int i;
+	void (*jmp)(void);
+
+	__disable_irq();
+
+	for (i = 0; i < ARRAY_SIZE(NVIC->ICER); i++) {
+		NVIC->ICER[i] = 0xFFFFFFFF;
+		NVIC->ICPR[i] = 0xFFFFFFFF;
+	}
+
+	LL_SYSCFG_SetRemapMemory(LL_SYSCFG_REMAP_SYSTEMFLASH);
+
+	jmp = (void (*)(void))(void (*)(void))(*((uint32_t *)((bootloader + 4))));
+
+	/*
+	 * We need to clear a few things set by the Zephyr early startup code
+	 */
+	__set_CONTROL(0);
+#if !defined(CONFIG_CPU_CORTEX_M0)
+	__set_BASEPRI(0);
+#endif
+
+	__set_MSP(*(uint32_t *)bootloader);
+
+	__enable_irq();
+
+	jmp();
+
+	CODE_UNREACHABLE;
+}
+
+static int bootloader_check_boot_init(void)
+{
+	if (bootmode_check(BOOT_MODE_TYPE_BOOTLOADER) > 0) {
+		bootmode_clear();
+		jump_to_bootloader();
+	}
+
+	return 0;
+}
+
+SYS_INIT(bootloader_check_boot_init, EARLY, 0);
```

Comment:
Won't you run into initialisation order issues here? The retention subsystem (and its backend drivers) are in post-kernel.

Reply:
So.... Yes, this is challenging. The stm32 bootloader, based on my research, can be persnickety if you try to jump to it with various things already init'd, like clocks, etc.

To be as robust as possible, I wanted to do this super early after a reset. The side effect is consuming device(s) that aren't technically in init'd yet, but we get *lucky* that the RAM based retained memory driver I'm using for my testing manages to work in this scenario. 

Some options:
* Move this stm32 bootloader code later, and enhance the code to properly reset/unload things before jumping.
* Move retention/retained memory bits earlier.
* Leave as is and it happens to work.

There's no real clear winner here, afaict

Reply:
> Some options:
> 
> * Move this stm32 bootloader code later, and enhance the code to properly reset/unload things before jumping.

That's certainly an option, but it might be complicated to deinitialize  everything.

> * Move retention/retained memory bits earlier.

I think this would be a good solution, but I doubt we can move it all the way to the `EARLY` initialisation level. @nordicjm: Any thoughts on this one?

> * Leave as is and it happens to work.

This is not an option. We cannot disregard the whole device model and initialisation order and rely on something that just happens to work.

Reply:
It should not be moved to earlier than `POST_KERNEL` to allow for drivers to init, e.g. it might need i2c. There already is driver deinit functionality and some drivers have it so makes sense to use that

Reply:
So far, the *cleanest*, but perhaps annoying to get right for all platforms, would be to move this init hook to a later level/priority that is correct from a sequence standpoint, and see what de-init is really needed. I will experiment with this.

If *device PM* is enabled, we could potentially leverage that to shut down any devices that are already loaded? Lacking that, I'm not sure how best to do this generically without just making the STM32 LL calls ourselves like `LL_RCC_DeInit`, etc.